source/s3_util.c

```diff
@@ -307,3 +314,161 @@ void aws_s3_add_user_agent_header(struct aws_allocator *allocator, struct aws_ht
     /* Clean up the scratch buffer. */
     aws_byte_buf_clean_up(&user_agent_buffer);
 }
+
+int aws_s3_parse_content_range_response_header(
+    struct aws_allocator *allocator,
+    struct aws_http_headers *response_headers,
+    uint64_t *out_range_start,
+    uint64_t *out_range_end,
+    uint64_t *out_object_size) {
+    AWS_PRECONDITION(allocator);
+    AWS_PRECONDITION(response_headers);
+    AWS_PRECONDITION(out_object_size);
+
+    struct aws_byte_cursor content_range_header_value;
+
+    if (aws_http_headers_get(response_headers, g_content_range_header_name, &content_range_header_value)) {
+        aws_raise_error(AWS_ERROR_S3_MISSING_CONTENT_RANGE_HEADER);
+        return AWS_OP_ERR;
+    }
+
+    int result = AWS_OP_ERR;
+
+    uint64_t range_start = 0;
+    uint64_t range_end = 0;
+    uint64_t object_size = 0;
+
+    struct aws_string *content_range_header_value_str =
+        aws_string_new_from_cursor(allocator, &content_range_header_value);
+
+    /* Expected Format of header is: "bytes StartByte-EndByte/TotalObjectSize" */
+    int num_fields_found = sscanf(
+        (const char *)content_range_header_value_str->bytes,
+        "bytes %" PRIu64 "-%" PRIu64 "/%" PRIu64,
+        &range_start,
+        &range_end,
+        &object_size);
+
+    if (num_fields_found < 3) {
+        aws_raise_error(AWS_ERROR_S3_INVALID_CONTENT_RANGE_HEADER);
+        goto clean_up;
+    }
+
+    if (out_range_start != NULL) {
+        *out_range_start = range_start;
+    }
+
+    if (out_range_end != NULL) {
+        *out_range_end = range_end;
+    }
+
+    *out_object_size = object_size;
+    result = AWS_OP_SUCCESS;
+
+clean_up:
+    aws_string_destroy(content_range_header_value_str);
+    content_range_header_value_str = NULL;
+
+    return result;
+}
+
+/* Parse the content-length from a content-length respone header.*/
+int aws_s3_parse_content_length_response_header(
+    struct aws_allocator *allocator,
+    struct aws_http_headers *response_headers,
+    uint64_t *out_content_length) {
+    AWS_PRECONDITION(allocator);
+    AWS_PRECONDITION(response_headers);
+    AWS_PRECONDITION(out_content_length);
+
+    struct aws_byte_cursor content_length_header_value;
+
+    if (aws_http_headers_get(response_headers, g_content_length_header_name, &content_length_header_value)) {
+        aws_raise_error(AWS_ERROR_S3_MISSING_CONTENT_LENGTH_HEADER);
+        return AWS_OP_ERR;
+    }
+
+    struct aws_string *content_length_header_value_str =
+        aws_string_new_from_cursor(allocator, &content_length_header_value);
+    char *content_length_str_end = NULL;
+
+    *out_content_length = strtoull((const char *)content_length_header_value_str->bytes, &content_length_str_end, 10);
```

Comment:
Error check here? Check the `content_length_str_end == '\0'`? And referring to a stack overflow [answer](https://stackoverflow.com/a/11279793), why not use `sscanf` instead like the previous function?

Reply:
Using sscanf, added error check.